### ISSUE TITLE
Optional spline interpolation on shapes

### ISSUE THREAD

**Issue by half-adder** [napari/napari#1256]
## 🚀 Feature

Shapes could have the option of a smooth interpolation between points.

## Motivation
In my image analysis pipeline, I make use of smooth curves which interpolate through points. This functionality is also available in e.g. FIJI/ImageJ.

## Pitch

My initial idea is that in addition to the opacity/edge width sliders and the blending/face/edge color dropdowns, there could be a "smooth" checkbox which would enable cubic interpolation when checked, and linear interpolation when unchecked. Alternatively,  there could be a dropdown menu with different interpolation options.

## Alternatives

These smooth versions could be their own shapes? That seems less elegant though.

## Additional context

This interpolation is actually quite easy thanks to SciPy. I made a rough proof-of-concept by calling `_set_meshes` with the interpolated data within `_update_displayed_data` in `napari/layers/shapes/_shapes_models/path.py`:

```python
def _update_displayed_data(self):
        """Update the data that is to be displayed."""
        # For path connect every all data

        if self.data_displayed.shape[0] > 2:
            distance = np.cumsum(np.sqrt(np.sum(np.diff(self.data_displayed[:-1,:], axis=0)**2, axis=1 )) )
            distance = np.insert(distance, 0, 0)/distance[-1]

            # the number of sampled data points might need to be carefully thought
            # about (might need to change with image scale?)
            alpha = np.linspace(0, 1, 75)

            spl = CubicSpline(distance, self.data_displayed[:-1, :])
            interp_pts = spl(alpha)

            self._set_meshes(interp_pts, face=False, closed=False)
        else:
            self._set_meshes(self.data_displayed, face=False, closed=False)

        self._box = create_box(self.data_displayed)

        data_not_displayed = self.data[:, self.dims_not_displayed]
        self.slice_key = np.round(
            [
                np.min(data_not_displayed, axis=0),
                np.max(data_not_displayed, axis=0),
            ]
        ).astype('int')
```

I recorded a short video of this in action
![napari_spline](https://user-images.githubusercontent.com/10676434/82109899-7c08cd80-96ff-11ea-9d0a-d508eacbc98a.gif)

I'm not really familiar with napari's architecture, so this is all I've done so far, but if the above proposal sounds interesting, I'd be happy to keep working on this.

**Comment by sofroniewn** [comment]
Wow @half-adder this is beautiful!! So amazing to see a gif with a proof of concept already. We'd love to have this functionality get added into napari, and I'm happy to work with you to try and get it in.

There are some UI things we'll need to think about, and also we'll need to maybe watch out for any performance related problems that could come up if we have many lines that are getting smoothed.

A checkbox for "smooth" seems like a good place to start.

I'm wondering if people will very quickly want additional features and UI elements to control the details of the smoothing, like in Illustrator where you have additional handles to adjust curvature. I don't think we want to start supporting that, as I see that as moving away from scientific image visualization and quantification and more towards design and graphics, but I'm curious if people would have a need in which case we'd discuss it more.

![image](https://user-images.githubusercontent.com/6531703/82126239-57910d80-9760-11ea-8fb7-7ee00cd14059.png)

It would probably be nice if the checkbox is applied at the level of individual shapes, so that you could have one layer with three shapes, one which has smoothing, two that don't.

We might also want to expose "number of sampled data points" as a parameter on the UI (which if am correct will control degree of smoothing to some extent right?) rather than hard code it. It might also be nice to let that be controlled on the individual shape level.

Curious how @jni thinks about all this, I don't know if he uses splines in his work/ has some ideas.

All in all though very exciting to see this!

**Comment by tlambert03** [comment]
I love this too.  very cool.

I'm also curious how hard it will be to extract data along such a spline curve?  For instance, to get the intensity in an image layer along the curve of some shape layer, or an average intensity within the bounds of some curvy shape.  Beyond displaying these curves, any thoughts on how deep we need to go to make good use from them?  @sofroniewn, @half-adder?

**Comment by half-adder** [comment]
Excited that this has some interest!

@tlambert03 I do have code that can measure along these curves - I use them to make measurements in my microscopy. [Here's the function I'm using to do it](https://github.com/ApfeldLab/pharynx_redox/blob/39a2f955e8ff8445f3809603df02f40947257433/pharynx_redox/image_processing.py#L547).

Basically i just use `np.map_coordinates` to sample the image along the points of the spline.

If you want the avg. intensity around the shape, it does get a bit more complex (although there might be a much easier way than what I'm doing!). I think if the shape has been given width through the UI it makes sense to do.

For this, I just calculate the normal lines at each sampled point along the curve, then use `np.map_coordinates` to measure under those, resulting in a measurement rectangle (which is essentially a "straightened" image along the curve). Then you can flatten it with the average. In that sample code, I actually weight the average around a gaussian centered on the curve, but that is specific to my needs, and I'm guessing is needlessly complex for this package.

**Comment by jni** [comment]
> Curious how @jni thinks about all this, I don't know if he uses splines in his work/ has some ideas.

I do not, but I agree with everything that's been said here so far. =) I agree that we have to be careful not to become illustrator vs an analysis tool, but other than that this looks great and useful.

Things like additional control points can come in later PRs.

Thanks @half-adder!

**Comment by sofroniewn** [comment]
There was an attempt to do this in #2345 which could be picked up by folks. We're certainly not far from this functionality, but I don't have bandwidth to finish it up right now so am closing that PR.

**Comment by sofroniewn** [comment]
See here for an explanation of what might be required to finish the work https://github.com/napari/napari/pull/2345#issuecomment-955157371

**Comment by dstansby** [comment]
Is there still work to do on this after https://github.com/napari/napari/pull/5334?

**Comment by psobolewskiPhD** [comment]
I think exposing in the UI would be nice—after all, drawing shapes is a GUI feature.

**Comment by brisvag** [comment]
Oof, this is still hanging, sorry :/

Yeah it's not directly exposed in the API, but it should be pretty easy. Maybe it's time to do that :) @melonora @psobolewskiPhD you touched shapes gui code recently, would you like to give it a go and/or have opinions about how this should be exposed? I think the main issue was whether we want all shapes in a layer to be affected or if we want some finer control.

**Comment by melonora** [comment]
I would like to give it a go, but perhaps better to discuss gui interface in a community meeting first?

**Comment by psobolewskiPhD** [comment]
I think step 1 needs to be to sort out programmatic access so we can understand the behavior. Then once that is clear, can implement a GUI.
Also, will this have any repercussions on saving Shapes layers?

**Comment by brisvag** [comment]
The way interpolation works here is only on the visualization side. The data itself remains unchanged. So anything that uses the layer data will be unaffected, unless it reaches into the depths of the `Polygon` or `vispy` objects and gets data from there.

**Comment by melonora** [comment]
yes, but I do see it critical that interpolation information is actually saved alongside the data. Otherwise there would be no way for 2 people to recreate the view or perhaps rasterize for analysis purposes / doing certain measurements such as distances.

**Comment by brisvag** [comment]
Yep, but my gut feeling is that ideally we would save the interpolation parameters rather than the actual output?

**Comment by melonora** [comment]
Oh I wouldn't go with the actual output of interpolation within the shape itself, but for example geopandas has a `.geometry` column containing the geometry, but then another column can contain per shape the kind of interpolation for example. This allows for creating the actual output for analysis purposes when needed.

**Comment by carlosmariorr** [comment]
I'd love to bring this up in a community meeting next week. I'll make a pitch of my ideas. If anyone is still working on this I'd love to help out or pick it up if it's dropped.
